2140 time at noon and cache

glintik opened the pull request:
Two commits that fix #2140:
1) Set time to noon(12:00) for cloned Moment objects with startOf() calls in DayPickerRangeController. 
2) Added cache for isDayBlocked and isOutsideRange.

ljharb commented on src/components/DayPickerRangeController.jsx:
```diff
@@ -213,11 +213,12 @@ export default class DayPickerRangeController extends React.PureComponent {
 
     this.isTouchDevice = isTouchDevice();
     this.today = moment();
+    this.cachedProps = {};
```
This will break use cases; it’s not guaranteed that these callbacks’ return values are memoizable.

glintik replied:
I think, these callbacks return values can be memoizable during render time at least, but it's up to you to decide.

ljharb replied:
During a single render, that seems reasonable, but also I’m not sure what the value is of adding this optimization.

glintik replied:
Value is performance improvement. 3x less calls of isDayBlocked() and isOutsideRange(). If user has DayPickerRangeController with both callbacks used, then 90(3 months)*2(methods)*3(not cached) = 540 calls performed. After the improvement with return values cache, calls count decreases to 180.

ljharb replied:
hmm. this seems like a bit more complexity than should be necessary for per-render caching. could we split that out into a separate PR, and keep this one just for the bug fix and regression tests?

glintik replied:
Moved cache commit to PR #2147

ljharb commented:
For the first commit, can we add regression tests?

glintik commented:
> For the first commit, can we add regression tests?

Yes, I can add some. You mean, to check time in changed methods or something more sophisticated?

ljharb commented:
Something that replicates the bug, and that your changes fix, without directly testing midnight vs noon.

glintik commented:
@ljharb, please take a look again. I reverted cache changes and added some tests.
P.S. I wasn't able to run tests from DayPicker_spec.jsx, looks like tests are broken?
<img width="890" alt="Screenshot 2021-09-21 at 14 29 10" src="https://user-images.githubusercontent.com/2947475/134162919-3ec1a515-46d9-433a-865f-5c67394470a9.png">

ljharb commented on test/components/DayPickerSingleDateController_spec.jsx:
```diff
@@ -974,6 +974,19 @@ describe('DayPickerSingleDateController', () => {
       expect(firstFocusableDay.isSame(date, 'day')).to.equal(true);
     });
 
+    it('time is a noon', () => {
+      sinon.stub(DayPickerSingleDateController.prototype, 'isBlocked').returns(false);
+      const wrapper = shallow((
+        <DayPickerSingleDateController
+          date={null}
+          onDateChange={sinon.stub()}
+          onFocusChange={sinon.stub()}
+        />
+      ));
+      const firstFocusableDay = wrapper.instance().getFirstFocusableDay(today);
+      expect(firstFocusableDay.hours()).to.equal(12);
+    });
```
this seems like it's testing the implementation, rather than the bug reported in #2140.

glintik replied:
@ljharb, not sure I follow what you mean.
From bug:
> expected behavior: returns consistent time, either always midnight local time or always noon local time
Test ensures that time is a noon. Bug closed.

ljharb replied:
right, but the bug is "it's called twice and the time isn't the same" - so shouldn't the test call it twice and guarantee the time is the same?

(it's fine to *also* test it's at noon, because there's other reasons it should be noon)

glintik replied:
"called twice" resolved in another commit, I sent you another PR with cache solution to resolve it. This PR fixes time issue only.

ljharb replied:
The issue was that *when* it's called multiple times, even across renders, the *time* is different. I don't think the caching would address that.

glintik replied:
@ljharb As I said, no caching involved in my solution in current PR. I reverted it. The only changes - to set up hours to 12:00.

ljharb reviewed (changes requested):
Thanks! I've rebased this.

There's still changes in `utils/getNumberOfCalendarMonthWeeks` and `utils/isDayVisible` and `components/DayPicker` that don't have associated tests, though.

glintik commented:
> Thanks! I've rebased this.
> 
> There's still changes in `utils/getNumberOfCalendarMonthWeeks` and `utils/isDayVisible` and `components/DayPicker` that don't have associated tests, though.

hi @ljharb,

May you refer to an example, what test needed for these changes?

ljharb commented on src/utils/getNumberOfCalendarMonthWeeks.js:
```diff
@@ -9,7 +9,7 @@ export default function getNumberOfCalendarMonthWeeks(
   firstDayOfWeek = moment.localeData().firstDayOfWeek(),
 ) {
-  const firstDayOfMonth = month.clone().startOf('month');
+  const firstDayOfMonth = month.clone().startOf('month').hour(12);
```
i assume that `getNumberOfCalendarMonthWeeks` has a bug in some timezones if the hour is set to 0 - can we have a unit test that triggers that bug, such that this line's change fixes it?

glintik replied:
To be honest, getNumberOfCalendarMonthWeeks() is not affected by the issue, I added "hour(12)" as unified approach to get firstDayOfMonth. Probably it's a good idea to have special util methods like :
```js
const INITIAL_HOUR = 12;

getFirstDayOfMonth = (month) => month.clone().startOf('month').hour(INITIAL_HOUR);
getFirstDayOfFirstWeekOfMonth = (month) => getFirstDayOfMonth(month).startOf('week').hour(INITIAL_HOUR);
```

ljharb replied:
Is there no timezone/date combination where the first day of the month, or the first day of the first week of a month, doesn't have a midnight?

glintik replied:
Midnight is 00:00, but I set hour(12) that is a noon to mitigate summer/winter time changes, so noon is a good choice for date calculations. Or you mean something other, sorry?

ljharb replied:
Right - I’m envisioning a bug caused by midnight that noon fixes.

glintik replied:
I did quick research and didn't find any other issues related to midnight, but react-dates codebase is large and no warranty here without full code audit.

ljharb replied:
What I mean is, 2013.10.20 was the start of the week in Brazil where midnight doesn't exist, but it's not the start of the month or the first day of the week when the week ends in a different month. Maybe there's another date in a specific timezone where either of those is true?

glintik replied:
> where midnight doesn't exist
I don't see that lack of midnight causes any error:
```js
const moment = require('moment-timezone');
moment.tz("2013-10-20 00:00:00", "Brazil/East").toString();
```
Output:
```Sun Oct 20 2013 01:00:00 GMT-0200```
Moment handles this flawlessly as I can see(+1 hour added).

Thank you for merging!

ljharb commented on src/utils/isDayVisible.js:
```diff
@@ -21,7 +21,7 @@ export default function isDayVisible(day, month, numberOfMonths, enableOutsideDa
 
   if (enableOutsideDays) {
     if (!startCacheOutsideDays.has(startKey)) {
-      startCacheOutsideDays.set(startKey, month.clone().startOf('month').startOf('week'));
+      startCacheOutsideDays.set(startKey, month.clone().startOf('month').startOf('week').hour(12));
```
same question here

glintik replied:
Same as for ```getNumberOfCalendarMonthWeeks```, just to honor unified approach to get first day of week/month.

ljharb commented on src/components/DayPicker.jsx:
```diff
@@ -695,7 +695,7 @@ class DayPicker extends React.PureComponent {
 
     if (newMonth && (!focusedDate || !isDayVisible(focusedDate, newMonth, numberOfMonths))) {
-      focusedDate = newMonth.clone().startOf('month');
+      focusedDate = newMonth.clone().startOf('month').hour(12);
```
similarly, something that unit tests or exercises getFocusedDay, and shows the bug that this change fixes, would cover here

glintik replied:
Bug is not related to timezone, it's related to initial time for newly created date objects because in some code lines hours are set while in others are not. So I'm in stuck what test is needed here.

ljharb replied:
If this change isn't fixing anything, maybe it should be reverted?

glintik replied:
> If this change isn't fixing anything, maybe it should be reverted?

Sure, but there is a good reason to have unified approach everywhere: it minimizes possible FUTURE bugs. All developers will see how to get first day of month and in all future cases will get a date with proper initial time.

Only "hour(12)" in components really matter for the issue.

ljharb commented:
I've played with a few of the dates here: https://stackoverflow.com/a/61328380/632724 and i can't come up with any test cases where the hour matters. I'll merge this as-is.
